### PR TITLE
Implement deploy command

### PR DESCRIPTION
**Pull request by runesoerensen**

### CODE REVIEW

File: src/AppHarbor/Commands/DeployAppCommand.cs [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+			var presignedUrl = client.Execute(urlRequest).Content;
+
+			var sourceDirectory = new DirectoryInfo(Directory.GetCurrentDirectory());
+
+			HttpWebRequest httpRequest = WebRequest.Create(presignedUrl) as HttpWebRequest;
+			httpRequest.Method = "PUT";
+			httpRequest.AllowWriteStreamBuffering = false;
+
+			var timeout = 10000000;
```
Comment (tdctaz) [inline src/AppHarbor/Commands/DeployAppCommand.cs]: Which unit is this, use TimeSpan.FromMinutes(xxx).TotalSeconds, or TotalMilliconds, or something.

File: src/AppHarbor/Commands/DeployAppCommand.cs [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+		public DeployAppCommand(IApplicationConfiguration applicationConfiguration, IAccessTokenConfiguration accessTokenConfiguration, TextReader reader, TextWriter writer)
+		{
+			_accessToken = accessTokenConfiguration.GetAccessToken();
+			_applicationConfiguration = applicationConfiguration;
+			_reader = reader;
+			_writer = writer;
+		}
+
+		public void Execute(string[] arguments)
```
Comment (tdctaz) [inline src/AppHarbor/Commands/DeployAppCommand.cs:29]: I think that I might split this method up into two parts, `PerformUpload(..., Action<ProgressStatus> progressCallback)` and then have the UI update kept separate in that callback.

File: src/AppHarbor/Commands/DeployAppCommand.cs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+			{
+				_writer.WriteLine("Package upload failed. Please try again.");
+				return;
+			}
+
+			_writer.WriteLine("Package successfully uploaded.");
+			TriggerAppHarborBuild(_applicationConfiguration.GetApplicationId(), presignedUrl);
+		}
+
+		private bool TriggerAppHarborBuild(string application_slug, string download_url)
```
Comment (tdctaz) [inline src/AppHarbor/Commands/DeployAppCommand.cs]: 'applicationSlug' and 'downloadUrl'

File: src/AppHarbor/Commands/DeployAppCommand.cs [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+				_writer.WriteLine("AppHarbor notified, deploying... Open overview in browser with `appharbor open`.");
+			}
+			return true;
+		}
+
+		private static void RenderConsoleProgress(int percentage, char progressBarCharacter, ConsoleColor color, string message)
+		{
+			Console.CursorVisible = false;
+			ConsoleColor originalColor = Console.ForegroundColor;
+			Console.ForegroundColor = color;
```
Comment (tdctaz) [inline src/AppHarbor/Commands/DeployAppCommand.cs]: try / finally around the set / restore ForegroundColor - any exceptions here will not reset the color back to its original value

File: src/AppHarbor/CompressionExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+using System.Linq;
+using ICSharpCode.SharpZipLib.Tar;
+
+namespace AppHarbor
+{
+	public static class CompressionExtensions
+	{
+		public static void ToTar(this DirectoryInfo sourceDirectory, Stream output)
+		{
+			var archive = TarArchive.CreateOutputTarArchive(output);
```
Comment (tdctaz) [inline src/AppHarbor/CompressionExtensions.cs:12]: Are you sure this library works, I know their tar library failed horrible with extraction. Can it archive long filename (100 char+) and localized characters?

File: src/AppHarbor/Commands/DeployAppCommand.cs [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+			httpRequest.Method = "PUT";
+			httpRequest.AllowWriteStreamBuffering = false;
+
+			var timeout = 10000000;
+			httpRequest.Timeout = timeout;
+			httpRequest.ReadWriteTimeout = timeout;
+
+			using (var memoryStream = new MemoryStream())
+			{
+				using (var gzipStream = new GZipStream(memoryStream, CompressionMode.Compress, true))
```
Comment (tdctaz) [inline src/AppHarbor/Commands/DeployAppCommand.cs]: GZipStream from Microsoft has random deadlock issues, they are rare so we could ignore it in the console tool.

File: src/AppHarbor/CompressionExtensions.cs (hunk 5 above)
Comment (runesoerensen) [inline src/AppHarbor/CompressionExtensions.cs:12]: It's the one we're currently using in the GitDownloadController so I assume it's been tested thoroughly with real apps. The older implementation didn't work well though.

File: src/AppHarbor/Commands/DeployAppCommand.cs (hunk 2 above)
Comment (runesoerensen) [inline src/AppHarbor/Commands/DeployAppCommand.cs:29]: I've moved the upload logic to it's own method, but I'm not sure how to implement the `Action<ProgressStatus> progressCallback` -  although it sounds like a good idea. Any recommendations?

File: src/AppHarbor/CompressionExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+			archive.Close();
+		}
+
+		private static IEnumerable<FileInfo> GetFiles(DirectoryInfo directory, string[] excludedDirectories)
+		{
+			IEnumerable<FileInfo> files = directory.GetFiles("*", SearchOption.TopDirectoryOnly);
+			foreach (var nestedDirectory in directory.GetDirectories())
+			{
+				if (excludedDirectories.Contains(nestedDirectory.Name))
```
Comment (tt) [inline src/AppHarbor/CompressionExtensions.cs]: This could be converted to a LINQ query.

File: src/AppHarbor/Commands/OpenAppCommand.cs [hunk 8]
```diff
@@ -0,0 +1,20 @@
+﻿using System.Diagnostics;
+
+namespace AppHarbor.Commands
+{
+	[CommandHelp("Open application on appharbor.com", alias: "open")]
+	public class OpenAppCommand : ICommand
```
Comment (tt) [inline src/AppHarbor/Commands/OpenAppCommand.cs:6]: `OpenApplicationCommand`

File: src/AppHarbor/Commands/DeployAppCommand.cs [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+			Console.CursorLeft = 0;
+
+			try
+			{
+				Console.CursorVisible = false;
+				Console.ForegroundColor = color;
+
+				int width = Console.WindowWidth - 1;
+				int newWidth = (int)((width * percentage) / 100d);
+				string progressBar = new string(progressBarCharacter, newWidth) +
```
Comment (tt) [inline src/AppHarbor/Commands/DeployAppCommand.cs]: Use `PadRight`.

File: src/AppHarbor/Commands/DeployAppCommand.cs [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+			_writer.WriteLine("The package will be deployed to application \"{0}\".", _applicationConfiguration.GetApplicationId());
+			_writer.Write("Enter a deployment message:");
+			var commitMessage = _reader.ReadLine();
+
+			var request = new RestRequest("applications/{slug}/buildnotifications", Method.POST)
+			{
+				RequestFormat = DataFormat.Json
+			}
+				.AddUrlSegment("slug", applicationSlug)
+				.AddHeader("Authorization", "BEARER " + _accessToken)
```
Comment (tt) [inline src/AppHarbor/Commands/DeployAppCommand.cs]: `string.Format`

File: src/AppHarbor/Commands/DeployAppCommand.cs [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+			var response = client.Execute(request);
+
+			if (response.StatusCode == HttpStatusCode.OK)
+			{
+				_writer.WriteLine("AppHarbor notified, deploying... Open application overview with `appharbor open`.");
+			}
+			return true;
+		}
+
+		private void RenderConsoleProgress(int percentage, char progressBarCharacter, ConsoleColor color, string message)
```
Comment (tt) [inline src/AppHarbor/Commands/DeployAppCommand.cs]: I think the `percentage` should be a double to make for a smoother movement (not because you can draw half a character, but because the double rounding makes it jumpy).

Example (width of 40 characters):
- 1 % = 0.4: 0 characters
- 2 % = 0.8: 1 character
- 3 % = 1.2: 1 character
- 4 % = 1.6: 2 characters
- 5 % = 2.0: 2 characters
- 6 % = 2.4: 2 characters

(These values will not change, but allowing the input to contain decimals improve the overall calculation.)

File: src/AppHarbor/Commands/OpenAppCommand.cs (hunk 8 above)
Comment (runesoerensen) [inline src/AppHarbor/Commands/OpenAppCommand.cs:6]: The `OpenAppCommand` is in line with the existing solution (where command lines parameters are based on the class names). We might want to change that solution, but I think that'd be a separate pull request

File: src/AppHarbor/Commands/DeployAppCommand.cs [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+				var initiateMultipartUploadRequest = new InitiateMultipartUploadRequest
+				{
+					BucketName = uploadCredentials.Bucket,
+					Key = uploadCredentials.ObjectKey,
+				};
+
+				var initMultipartUploadResponse = s3Client.InitiateMultipartUpload(initiateMultipartUploadRequest);
+
+				var uploadResponses = new ConcurrentBag<UploadPartResponse>();
+				var waitHandlers = new List<WaitHandle>();
```
Comment (tt) [inline src/AppHarbor/Commands/DeployAppCommand.cs]: `waitHandles`

File: src/AppHarbor/Commands/DeployAppCommand.cs (hunk 2 above)
Comment (tdctaz) [inline src/AppHarbor/Commands/DeployAppCommand.cs:29]: I think it is fine, my idea was to get it out into its own class and then call with status updates at the appropriate locations (with some Begin, Update, End)

File: src/AppHarbor/Commands/DeployAppCommand.cs [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+		private FederatedUploadCredentials GetCredentials()
+		{
+			var urlRequest = new RestRequest("applications/{slug}/uploadCredentials", Method.POST);
+			urlRequest.AddUrlSegment("slug", _applicationConfiguration.GetApplicationId());
+
+			var federatedCredentials = _restClient.Execute<FederatedUploadCredentials>(urlRequest);
+			return federatedCredentials.Data;
+		}
+
+		private KeyValuePair<DateTime, double> lastUploadProgressEvent;
```
Comment (tdctaz) [inline src/AppHarbor/Commands/DeployAppCommand.cs]: missing _

File: src/AppHarbor/Commands/DeployAppCommand.cs [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+		private IList<double> bytesPerSecondAverages = new List<double>();
```
Comment (tdctaz) [inline src/AppHarbor/Commands/DeployAppCommand.cs]: missing _

File: src/AppHarbor/Commands/DeployAppCommand.cs [hunk 15]
```diff
@@ -0,0 +1,170 @@
...
+			using (var packageStream = new TemporaryFileStream())
+			{
+				using (var gzipStream = new GZipStream(packageStream, CompressionMode.Compress, true))
+				{
+					var sourceDirectory = new DirectoryInfo(Directory.GetCurrentDirectory());
+					sourceDirectory.ToTar(gzipStream, excludedDirectoryNames: new[] { ".git", ".hg" });
+
+					using (var s3Client = new AmazonS3Client(uploadCredentials.GetSessionCredentials()))
+					{
+						var uploadParts = 5;
```
Comment (tdctaz) [inline src/AppHarbor/Commands/DeployAppCommand.cs]: const int uploadParts = 5;

File: src/AppHarbor/Commands/DeployAppCommand.cs [hunk 16]
```diff
@@ -0,0 +1,170 @@
...
+				string.Format("{0} left", TimeSpan
+				.FromSeconds(bytesRemaining / WeightedAverage(bytesPerSecondAverages))
+				.GetHumanized());
+
+			ConsoleProgressBar.Render(uploadProgressArgs.PercentDone,
+				string.Format("Uploading package ({0}% of {1:0.0} MB). {2}",
+					uploadProgressArgs.PercentDone, uploadProgressArgs.TotalBytes / 1048576, timeEstimate));
+		}
+
+		private bool TriggerAppHarborBuild(string applicationSlug, FederatedUploadCredentials credentials)
```
Comment (tdctaz) [inline src/AppHarbor/Commands/DeployAppCommand.cs]: Change to void, is always true and not used

**Comment by runesoerensen** [comment]
Updated
